pan-domain-auth-play/src/main/scala/com/gu/pandomainauth/service/OAuth.scala

```diff
@@ -96,8 +97,8 @@ class OAuth(config: OAuthSettings, system: String, redirectUrl: String)(implicit
                   ),
                   authenticatingSystem = system,
                   authenticatedIn = Set(system),
-                  jwt.claims.exp * 1000,
-                  false
+                  expires = Instant.ofEpochSecond(jwt.claims.exp), // https://stackoverflow.com/a/39926886/438886
```

Comment:
On reading this through with fresh eyes, I actually think the seconds/milliseconds stuff is confusing enough that it warrants a more explicit comment.

My first question when seeing `ofEpochSecond` would be "wait, was this a mistake or intentional?", and I don't think linking to StackOverflow is enough to reassure the reader that the JWT/Panda distinction has been deliberately considered.

```suggestion
                  // The JWT standard specifies that `exp` is a `NumericDate`,
                  // which is defined as an epoch time in *seconds*
                  // (unlike the Panda cookie `expires` which is in milliseconds)
                  // https://www.rfc-editor.org/rfc/rfc7519#section-4.1.4
                  expires = Instant.ofEpochSecond(jwt.claims.exp),
```

Reply:
> I don't think linking to StackOverflow is enough to reassure the reader that the JWT/Panda distinction has been deliberately considered.

Ah, fair enough! You're right that there could always be doubt in someone's mind that converting from seconds is correct, when [elsewhere](https://github.com/guardian/pan-domain-authentication/blob/f03046ae00f9f60ae2b9b6526ccff68452161f2d/pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/service/CookieUtils.scala#L49) we've converted from milliseconds.

Personally tho, I avoid inline comments _within_ code blocks (I'm a _bit_ friendlier to comments against a class, method or field). They're prone to:

* **getting separated from the code they relate to** - many times in old codebases, I’ve seen lines of code deleted, _without_ their explanatory comment on the line above being deleted as well - leaving a confusing comment, apparently speaking to a different line of code further down.
* **becoming outdated** - they speak with an authority and _apparent_ currentness there's no reasonable mechanism for enforcing: code comments _look_ like they apply to the code _as it is right now_. In contrast documentation we write in PR descriptions/comments is understood in its context - when others come back to read it (which admittedly, assumes they've done a `git blame` and found the PR!), they understand the _age_ of that context: they understand that it was true when the PR was written, but might _not_ be now.

One complementary approach might be having a test that actually _exercises_ this code (we would probably want to refactor the construction of the `AuthenticatedUser` out into its own method for easier unit testing, and supply reference JSON for both the token and OAuth response) - we could explicitly assert that `expires` ends up taking a sensible value. If someone thinks `.ofEpochSecond(jwt.claims.exp)` is wrong, and tries to change it to `ofEpochMillis`, the test will fail, and there's no way they can make it pass without either editing the reference json or changing the expected `expires` value to a ridiculous value (thousands of years in the future!).

Incidentally, note that in the new comment, there's a statement that's vulnerable to becoming untrue over time:

```
// (unlike the Panda cookie `expires` which is in milliseconds)
```

...the format of the Panda cookie could change in future - it could even become a JWT, and be in seconds! If that were to happen, there's no test that would fail and say "Hey, there's now this comment in our codebase which is confusing/misleading!" - the comment would have to be manually spotted, and corrected.

Having said all that, I guess I just wanted to express my [personal](https://docs.google.com/document/d/19Ph05_rACj-EiX3GUCHuErAR_iy4kXN-NptOd8IAkDk/edit?usp=sharing) preferences on code documentation, and there are certainly arguments against them too. Please forgive me for banging on - and thank you for adding the clarification!